Fix generic parameter data flow validation in NativeAOT

vitek-karas opened the pull request:
This reworks how generic parameter data flow validation is done in the NativeAOT compiler.

Previously generic data flow was done from generic dictionary nodes. Problem with that approach is that there's no origin information at that point. The warnings can't point to the place where the problematic instantiation is in the code - we only know that it exists.
Aside from it being unfriendly for the users, it means any RUC or suppressions don't work on these warnings the same way they do in linker/analyzer.

This change modifies the logic to tag the method as "needs data flow" whenever we spot an instantiation of an annotated generic in it somewhere.
Then the actual validation/marking is done from data flow using the trim analysis patterns.

The only exception to this is generic data flow for base types and interface implementations, that one is done on the EEType nodes.

Note that AOT implements a much more precise version of the generic data flow validation as compared to linker/analyzer. See the big comment at the beginning of `GenericParameterWarningLocation.cs` for how that works.

Test changes:
Adds the two tests from linker which cover this functionality.

Change the test infra to use token to compare message origins for expected warnings. Consistently converting generic types/methods into strings across two type systems is just very difficult - the tokens are simple and reliable.

Changes the tests to avoid expecting specific generic types/methods formatting in the messages - again, it's too hard to make this consistent without lot of effort. And the tests don't really need it.

Fixes https://github.com/dotnet/runtime/issues/77455
Fixes https://github.com/dotnet/runtime/issues/75898

vitek-karas committed:
Implements generic parameter data flow mostly trhough data flow scanner

Previously generic data flow was done from generic dictionary nodes. Problem with that approach is that there's no origin information at that point. The warnings can't point to the place where the problematic instantiation is in the code - we only know that it exists.
Aside from it being unfriendly for the users, it means any RUC or suppressions don't work on these warnings the same way they do in linker/analyzer.

This change modifies the logic to tag the method as "needs data flow" whenever we spot an instantiation of an annotated generic in it somewhere.
Then the actualy validation/marking is done from data flow using the trim analysis patterns.

The only exception to this is generic data flow for base types and interface implementations, that one is done on the EEType nodes.

Note that AOT implements a much more precise version of the generic data flow validation as compared to linker/analyzer. See the big comment at the bening of GenericParameterWarningLocation.cs for how that works.

vitek-karas committed:
Fix build after rebase

vitek-karas committed:
Add GenericParameterDataFlow tests - change test infra

Changes the expected warning validation to use tokens to compare message origins (same reason as with Kept validation - consistently converting things to string is hard)

sbomer reviewed (approved):
Looks great, thank you! I spent the most time on the parts I am more familiar with, so you might still want to get another review.

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/FlowAnnotations.cs:
```diff
@@ -105,6 +106,31 @@ public bool HasAnyAnnotations(TypeDesc type)
+            {
+                return GetAnnotations(type.GetTypeDefinition()).HasGenericParameterAnnotation();
+            }
+            catch (TypeSystemException)
```
It's when we hit missing/unresolvable dependencies, basically.

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/UsageBasedMetadataManager.cs:
```diff
@@ -403,6 +403,38 @@ private static bool IsTrimmableAssembly(ModuleDesc assembly)
+                    GetDataFlowDependenciesForInstantiation(ref dependencies, factory, type.BaseType, type);
+                }
+
+                foreach (var runtimeInterface in type.RuntimeInterfaces)
+                {
+                    GetDataFlowDependenciesForInstantiation(ref dependencies, factory, runtimeInterface, type);
+                }
```
This goes over all interfaces, including those that are inherited from the bases. What's the illinker behavior for.

```
interface IFoo<[DAM(Methods)] T> { }

[Suppress(The warning that Foo's T doesn't have DAM)]
class Foo<T> : IFoo<T> { }

// Does the suppression on Foo apply here? Bar's RuntimeInterfaces includes IFoo
class Bar<T> : Foo<T> { }
```

(This is obviously in the "obscure corner cases very few people will care about" bucket).

vitek-karas replied:
Good catch - what would be the API to get just the declared interfaces?
It's definitely a corner case, because Roslyn typically flattens out this during compilation anyway, but still.

MichalStrehovsky replied:
There is `.ExplicitlyImplementedInterfaces` but IIRC because of Roslyn's flattening, `Bar<T>` would still have `IFoo` there even though nobody put it there. We would have to look at interfaces of the base type and whether the problem was introduced in the base type to distinguish it from `class YetAnotherBar<T> : IFoo<T>, Foo<T> { }`. Maybe just leave as wart?

vitek-karas replied:
I'll look at what linker does (I assume it just reads the IL so will see the same flattening) and probably leave it as is if it does the same.

vitek-karas replied:
I changed it to go over the explicitly implemented ones only.

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/UsageBasedMetadataManager.cs:
```diff
@@ -776,69 +791,15 @@ public override DependencyList GetDependenciesForCustomAttribute(NodeFactory fac
             }
         }
 
-        public override void GetDependenciesForGenericDictionary(ref DependencyList dependencies, NodeFactory factory, TypeDesc type)
```
Can we delete the base virtual method too? AFAIK it's empty.

vitek-karas replied:
We can, I didn't know what is the typical solution - if we want to keep it because it's called from the right places, or just remove it.

MichalStrehovsky replied:
It was introduced specifically for this. We can reintroduce again if needed.

MichalStrehovsky reviewed (approved):
Looks great otherwise!

vitek-karas commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/UsageBasedMetadataManager.cs:
```diff
@@ -403,6 +403,38 @@ private static bool IsTrimmableAssembly(ModuleDesc assembly)
+        {
+            if (type.HasInstantiation && FlowAnnotations.HasGenericParameterAnnotation(type))
+            {
+                TypeDesc instantiatedType = type.InstantiateSignature(contextType.Instantiation, Instantiation.Empty);
```
I'll look into this, but it's probably a left over before I decided to use tokens in the tests. Without this sometimes the type was reported as
```
SomeGeneric`1
```
While what we want is something like
```
SomeGeneric<T>
```

vitek-karas replied:
This is actually necessary - it turns `BaseType<!0>` into `BaseType<TOuter>`. Otherwise we would have to map the signature variables to the derived type's generic parameters manually.

But you comment brought up a good point, that this logic was actually running potentially multiple times per type (for each instantiation) which could lead to duplicate or misleading warnings. So I rewrote this into a new dependency node and only run the analysis on the type definition (instantiations are just confusing, because they would have the base type also instantiated which could lead to misleading warnings).

From what I saw it's a rule that for each instantiated EEType there will be an EEType for the definition as well, but it felt relatively fragile to depend on it (I could not find it in the code written as a hard rule) - so I added a new dependency node which is only create for the type definition (and only if the type is interesting in some way).

vitek-karas committed:
PR feedback

The biggest change is to rework how the type inheritance checks are triggered. The new solution adds a new dependency node with the generic type definition to the graph and then does analysis on that node.
This is to avoid potential noise warnings which could happen due to multiple instantiations calling the checking code multiple times. With this the check is done only once on the type definition.

Tweaked some tests to try to cover the multiple instances scenario.

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/NodeFactory.cs:
```diff
@@ -704,6 +709,14 @@ public DataflowAnalyzedMethodNode DataflowAnalyzedMethod(MethodIL methodIL)
+
+        public DataflowAnalyzedTypeDefinitionNode DataflowAnalyzedTypeDefinition(TypeDesc type)
+        {
+            Debug.Assert(type.IsTypeDefinition);
```
Nit: we check this in the node constructor - it's the more appropriate place to check node invariants.

vitek-karas committed:
PR feedback

vitek-karas commented:
This change has been reverted in https://github.com/dotnet/runtime/pull/81259 due to #81358.
